slp2mp4/slp-to-mp4.py

```diff
@@ -40,10 +41,91 @@
 else:
     THIS_CONFIG = os.path.join(SCRIPT_DIR, 'config.json')
 THIS_USER_DIR = os.path.join(SCRIPT_DIR, 'User')
+OUT_DIR = os.path.join(SCRIPT_DIR, 'out')
+
+
+combined_files = []
+
 
 def is_game_too_short(num_frames, remove_short):
     return num_frames < MIN_GAME_LENGTH and remove_short
 
+
+def record_file_slp(slp_file, outfile):
+    conf = Config()
+
+    # Parse file with py-slippi to determine number of frames
+    slippi_game = Game(slp_file)
+    num_frames = slippi_game.metadata.duration + DURATION_BUFFER
+
+    if is_game_too_short(num_frames, conf.remove_short):
+        print("Warning: Game is less than 30 seconds and won't be recorded. Override in config.")
+        return
+
+    # Dump frames
+    with DolphinRunner(conf, THIS_USER_DIR, SCRIPT_DIR, JOB_ID) as dolphin_runner:
+        video_file, audio_file = dolphin_runner.run(slp_file, num_frames)
+
+        # Encode
+        ffmpeg_runner = FfmpegRunner(conf.ffmpeg)
+        ffmpeg_runner.run(video_file, audio_file, outfile)
+
+        print('Created {}'.format(outfile))
+
+
+def combine(conf):
+    if not conf.combine:
+        return
+    for subdir, dirs, files in os.walk(OUT_DIR):
+        basedir = os.path.basename(subdir)
+        if os.path.exists(os.path.join(subdir, 'concat_file.txt')):
+            os.remove(os.path.join(subdir, 'concat_file.txt'))
+        file_count = 0
+        with open(os.path.join(subdir, 'concat_file.txt'), 'w+') as concat_file:
+            lines = []
+            for file in files:
+                try:
+                    if file.split('.')[-1] == 'mp4' and os.path.join(subdir, file) not in combined_files:
+                        file_count = file_count + 1
+                        lines.append("file \'" + os.path.join(subdir, file) + "\'" + "\n")
+                except Exception:
+                    pass
+            concat_file.writelines(lines)
+        if file_count > 0 and not os.path.exists(os.path.join(OUT_DIR, basedir) + '.mp4'):
+            ffmpeg_runner = FfmpegRunner(conf.ffmpeg)
+            ffmpeg_runner.combine(os.path.join(subdir, 'concat_file.txt'), os.path.join(OUT_DIR, basedir) + '.mp4')
+            combined_files.append(os.path.join(OUT_DIR, basedir) + '.mp4')
+            if os.path.exists(os.path.join(OUT_DIR, basedir)) and os.path.exists(os.path.join(OUT_DIR, basedir) + '.mp4'):
+                shutil.rmtree(os.path.join(OUT_DIR, basedir))
```

Comment:
I think this only works when all the slippi files are in leaf nodes of the directory tree, because os.walk will traverse top to bottom by default.
If you have a directory A with some slippi files, then a subdirectory B in that directory with more slippi files, this will combine the replayts in A, then delete A (along with B).
Is this correct?

Reply:
You are right. I think reversing the walk order from bottom to top fixes this.

Reply:
Actually, it won't :(
In the scenario I described above, first directory B would get combined and removed, creating B.mp4 in directory A.
Then directory A would get combined and removed, creating A.mp4 in the parent directory (and removing B.mp4). Right?

Reply:
I think you should specify what the directory structure should be in the documentation, and not deal with other cases.

Reply:
I think I understand what's going on here now. There are no nested subfolder in the out folder. There is only ever one layer of folders in the out folder, even if there are nested folders in the input slp folders. This is because it creates a directory in the out folder with the name of the immediate parent directory of the input slp file (see readme for example). It does not recreate the nested folder structure of the input.

Reply:
I missed your comment about providing documentation. Is what is in the readme clear and sufficient?

Reply:
Yep I checked again, docs seem fine